core/src/main/java/hudson/util/RingBufferLogHandler.java

```diff
@@ -86,13 +75,36 @@ public static int getDefaultRingBufferSize() {
     }
 
     @Override
-    public synchronized void publish(LogRecord record) {
-        int len = records.length;
-        records[(start + size) % len] = new LogRecordRef(record);
-        if (size == len) {
-            start = (start + 1) % len;
-        } else {
-            size++;
+    public void publish(LogRecord record) {
+        if (record == null) {
+            return;
+        }
+        if (JenkinsJVM.isJenkinsJVM() && record.getParameters() != null) {
+            try {
+                LogRecord clone = new LogRecord(record.getLevel(), dummyFormatter.formatMessage(record));
+                clone.setLoggerName(record.getLoggerName());
+                clone.setMillis(record.getMillis());
```

Comment:
Since `LogRecord#setMillis` is deprecated in Java 9+, and since we plan on requiring Java 11 very soon, I think it would be worthwhile to leave a comment reminding ourselves to update this usage in the future. Otherwise, we may very well forget to do this and continue using the deprecated method.

```suggestion
                clone.setMillis(record.getMillis()); // TODO Java 9+ clone.setInstance(record.getInstant());
```

Reply:
Indeed. Still on the fence about cloning, vs. just mutating the original `LogRecord`.